AADApplication - Includsion of AppRoles field causes errors when config is applied

### Description of the issue

AADApplication resources that include the AppRoles field throw the following error when applied to Azure:

`##[error][CannotDeleteOrUpdateEnabledEntitlement] : Permission (scope or role) cannot be deleted or updated unless disabled`

### Microsoft 365 DSC Version

1.24.1120.1

### Which workloads are affected

Azure Active Directory (Entra ID)

### The DSC configuration

```powershell
# https://microsoft365dsc.com/resources/azure-ad/AADApplication/
AADApplication 'TestApp1' {
    AvailableToOtherTenants = $False
    DisplayName             = 'TestApp1'
    Ensure                  = 'Present'
    IdentifierUris          = @()
    KnownClientApplications = @()
    IsFallbackPublicClient  = $False
    AppRoles                = @(
        MSFT_MicrosoftGraphappRole {
            AllowedMemberTypes = @('Application')
            Id = '354ab5bc-bc09-4ae9-bfb6-93c46e111918'
            IsEnabled = $True
            Origin = 'Application'
            Value = 'Application.Read.All'
            DisplayName = 'Test role 1'
            Description = 'This is an app role'
        }
    )
    Owners                  = @('johndoe@contoso.com')
    Permissions             = @(
        MSFT_AADApplicationPermission {
            Name                = 'User.Read'
            Type                = 'Delegated'
            SourceAPI           = 'Microsoft Graph'
            AdminConsentGranted = $False
        }
    )
    PublicClient            = $False
    ReplyURLs               = @()
    ApplicationId           = $ApplicationId
    TenantId                = $TenantId
    CertificateThumbprint   = $Thumbprint
}
```

### Verbose logs showing the problem

```powershell
2024-12-05T00:31:25.9068666Z VERBOSE: [MACHINE-ID]:
2024-12-05T00:31:25.9074804Z [[AADApplication]TestApp1::[AzureAD]AzureAD_Configuration] Updating existing AzureAD
2024-12-05T00:31:25.9080315Z Application {TestApp1} with values:
2024-12-05T00:31:25.9083491Z
2024-12-05T00:31:25.9086866Z
2024-12-05T00:31:25.9090117Z
2024-12-05T00:31:25.9096155Z Key                    Value
2024-12-05T00:31:25.9099282Z
2024-12-05T00:31:25.9105415Z ---                    -----
2024-12-05T00:31:25.9108314Z
2024-12-05T00:31:25.9114299Z DisplayName            TestApp1
2024-12-05T00:31:25.9117354Z
2024-12-05T00:31:25.9123450Z GroupMembershipClaims  None
2024-12-05T00:31:25.9126583Z
2024-12-05T00:31:25.9132707Z IsFallbackPublicClient False
2024-12-05T00:31:25.9135786Z
2024-12-05T00:31:25.9141876Z OptionalClaims         {accessToken}
2024-12-05T00:31:25.9145040Z
2024-12-05T00:31:25.9150949Z AppRoles               {System.Collections.Hashtable}
2024-12-05T00:31:25.9154140Z
2024-12-05T00:31:25.9160258Z ApplicationId          555937d7-700a-4d64-a61f-dff9b84f3ce1
2024-12-05T00:31:25.9163099Z
2024-12-05T00:31:25.9169061Z SignInAudience         AzureADMyOrg
2024-12-05T00:31:25.9172400Z
2024-12-05T00:31:25.9179538Z Verbose                True
2024-12-05T00:31:25.9961435Z VERBOSE: [MACHINE-ID]:
2024-12-05T00:31:25.9968325Z [[AADApplication]TestApp1::[AzureAD]AzureAD_Configuration] Performing the operation
2024-12-05T00:31:25.9974804Z "Update-MgApplication_UpdateExpanded" on target "Call remote 'PATCH /applications/{application-id}' operation".
2024-12-05T00:31:26.7668151Z ##[error][CannotDeleteOrUpdateEnabledEntitlement] : Permission (scope or role) cannot be deleted or updated unless disabled
first.
    + CategoryInfo          : InvalidOperation: ({ ApplicationId...phApplication }:) [], CimException
    + FullyQualifiedErrorId : CannotDeleteOrUpdateEnabledEntitlement,Microsoft.Graph.PowerShell.Cmdlets.UpdateMgApplic
   ation_UpdateExpanded
    + PSComputerName        : localhost
```

### Environment Information + PowerShell Version

```powershell
Name                           Value
----                           -----
PSVersion                      5.1.20348.2849
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.20348.2849
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,7 +4,7 @@
 
 * AADApplication
   * Changed logic to remove all permissions when an empty array is specified.
-    FIXES [#5524](https://github.com/microsoft/Microsoft365DSC/issues/5524)
+    FIXES [#5534](https://github.com/microsoft/Microsoft365DSC/issues/5534)
 * AADFeatureRolloutPolicy
   * Fixed policy retrieval
     FIXES [#5521](https://github.com/microsoft/Microsoft365DSC/issues/5521)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,9 @@
 
 # UNRELEASED
 
+* AADApplication
+  * Changed logic to remove all permissions when an empty array is specified.
+    FIXES [#5524](https://github.com/microsoft/Microsoft365DSC/issues/5524)
 * AADFeatureRolloutPolicy
   * Fixed policy retrieval
     FIXES [#5521](https://github.com/microsoft/Microsoft365DSC/issues/5521)

--- a/Modules/Microsoft365DSC/DSCResources/MSFT_AADApplication/MSFT_AADApplication.psm1
+++ b/Modules/Microsoft365DSC/DSCResources/MSFT_AADApplication/MSFT_AADApplication.psm1
@@ -975,72 +975,81 @@ function Set-TargetResource
         }
     }
 
-    if ($needToUpdatePermissions -and -not [System.String]::IsNullOrEmpty($Permissions) -and $Permissions.Length -gt 0)
+    if ($needToUpdatePermissions -and $null -ne $Permissions)
     {
         Write-Verbose -Message "Will update permissions for Azure AD Application {$($currentAADApp.DisplayName)}"
-        $allSourceAPIs = $Permissions.SourceAPI | Select-Object -Unique
-        $allRequiredAccess = @()
 
-        foreach ($sourceAPI in $allSourceAPIs)
+        if ($Permissions.Length -eq 0)
         {
-            Write-Verbose -Message "Adding permissions for API {$($sourceAPI)}"
-            $permissionsForcurrentAPI = $Permissions | Where-Object -FilterScript { $_.SourceAPI -eq $sourceAPI }
-            $apiPrincipal = Get-MgServicePrincipal -Filter "DisplayName eq '$($sourceAPI)'"
-            $currentAPIAccess = @{
-                ResourceAppId  = $apiPrincipal.AppId
-                ResourceAccess = @()
-            }
-            foreach ($permission in $permissionsForcurrentAPI)
+            Write-Verbose -Message "Desired set of permissions is empty, removing all permissions on the app."
+            $allRequiredAccess = @()
+        }
+        else
             {
-                if ($permission.Type -eq 'Delegated')
-                {
-                    $scope = $apiPrincipal.Oauth2PermissionScopes | Where-Object -FilterScript { $_.Value -eq $permission.Name }
-                    $scopeId = $null
-                    if ($null -eq $scope)
-                    {
-                        $ObjectGuid = [System.Guid]::empty
-                        if ([System.Guid]::TryParse($permission.Name, [System.Management.Automation.PSReference]$ObjectGuid))
-                        {
-                            $scopeId = $permission.Name
-                        }
-                    }
-                    else
-                    {
-                        $scopeId = $scope.Id
-                    }
-                    Write-Verbose -Message "Adding Delegated Permission {$($scopeId)}"
-                    $delPermission = @{
-                        Id   = $scopeId
-                        Type = 'Scope'
-                    }
-                    $currentAPIAccess.ResourceAccess += $delPermission
-                }
-                elseif ($permission.Type -eq 'AppOnly')
-                {
-                    $role = $apiPrincipal.AppRoles | Where-Object -FilterScript { $_.Value -eq $permission.Name }
-                    $roleId = $null
-                    if ($null -eq $role)
-                    {
-                        $ObjectGuid = [System.Guid]::empty
-                        if ([System.Guid]::TryParse($permission.Name, [System.Management.Automation.PSReference]$ObjectGuid))
-                        {
-                            $roleId = $permission.Name
-                        }
-                    }
-                    else
-                    {
-                        $roleId = $role.Id
-                    }
-                    $appPermission = @{
-                        Id   = $roleId
-                        Type = 'Role'
-                    }
-                    $currentAPIAccess.ResourceAccess += $appPermission
-                }
-            }
-            if ($null -ne $currentAPIAccess)
+            $allSourceAPIs = $Permissions.SourceAPI | Select-Object -Unique
+            $allRequiredAccess = @()
+
+            foreach ($sourceAPI in $allSourceAPIs)
             {
-                $allRequiredAccess += $currentAPIAccess
+                Write-Verbose -Message "Adding permissions for API {$($sourceAPI)}"
+                $permissionsForcurrentAPI = $Permissions | Where-Object -FilterScript { $_.SourceAPI -eq $sourceAPI }
+                $apiPrincipal = Get-MgServicePrincipal -Filter "DisplayName eq '$($sourceAPI)'"
+                $currentAPIAccess = @{
+                    ResourceAppId  = $apiPrincipal.AppId
+                    ResourceAccess = @()
+                }
+                foreach ($permission in $permissionsForcurrentAPI)
+                {
+                    if ($permission.Type -eq 'Delegated')
+                    {
+                        $scope = $apiPrincipal.Oauth2PermissionScopes | Where-Object -FilterScript { $_.Value -eq $permission.Name }
+                        $scopeId = $null
+                        if ($null -eq $scope)
+                        {
+                            $ObjectGuid = [System.Guid]::empty
+                            if ([System.Guid]::TryParse($permission.Name, [System.Management.Automation.PSReference]$ObjectGuid))
+                            {
+                                $scopeId = $permission.Name
+                            }
+                        }
+                        else
+                        {
+                            $scopeId = $scope.Id
+                        }
+                        Write-Verbose -Message "Adding Delegated Permission {$($scopeId)}"
+                        $delPermission = @{
+                            Id   = $scopeId
+                            Type = 'Scope'
+                        }
+                        $currentAPIAccess.ResourceAccess += $delPermission
+                    }
+                    elseif ($permission.Type -eq 'AppOnly')
+                    {
+                        $role = $apiPrincipal.AppRoles | Where-Object -FilterScript { $_.Value -eq $permission.Name }
+                        $roleId = $null
+                        if ($null -eq $role)
+                        {
+                            $ObjectGuid = [System.Guid]::empty
+                            if ([System.Guid]::TryParse($permission.Name, [System.Management.Automation.PSReference]$ObjectGuid))
+                            {
+                                $roleId = $permission.Name
+                            }
+                        }
+                        else
+                        {
+                            $roleId = $role.Id
+                        }
+                        $appPermission = @{
+                            Id   = $roleId
+                            Type = 'Role'
+                        }
+                        $currentAPIAccess.ResourceAccess += $appPermission
+                    }
+                }
+                if ($null -ne $currentAPIAccess)
+                {
+                    $allRequiredAccess += $currentAPIAccess
+                }
             }
         }
 
@@ -1298,9 +1307,15 @@ function Test-TargetResource
 
     $CurrentValues = Get-TargetResource @PSBoundParameters
 
-    if ($CurrentValues.Permissions.Length -gt 0 -and $null -ne $CurrentValues.Permissions.Name -and $Permissions.Name.Length -gt 0)
+    if ($CurrentValues.Permissions.Length -gt 0 -and `
+        $null -ne $CurrentValues.Permissions.Name)
     {
-        $permissionsDiff = Compare-Object -ReferenceObject ($CurrentValues.Permissions.Name) -DifferenceObject ($Permissions.Name)
+        $differenceObject = $Permissions.Name
+        if ($null -eq $differenceObject)
+        {
+            $differenceObject = @()
+        }
+        $permissionsDiff = Compare-Object -ReferenceObject ($CurrentValues.Permissions.Name) -DifferenceObject $differenceObject
         $driftedParams = @{}
         if ($null -ne $permissionsDiff)
         {

